File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    vector<vector<int>> graph;
    vector<int> state; // 0 = unvisited, 1 = visiting, 2 = processed
    
    bool hasCycle(int course) {
        if (state[course] == 1) return true;  // back-edge -> cycle
        if (state[course] == 2) return false; // already safe
        
        state[course] = 1;
        for (int next : graph[course]) {
            if (hasCycle(next)) return true;
        }
        state[course] = 2;
        
        return false;
    }
    
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        graph.assign(numCourses, {});
        state.assign(numCourses, 0);
        
        for (auto& p : prerequisites) {
            graph[p[1]].push_back(p[0]);
        }
        
        for (int i = 0; i < numCourses; i++) {
            if (state[i] == 0 && hasCycle(i)) {
                return false;
            }
        }
        
        return true;
    }
};

File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # Build adjacency list and in-degree count
        graph = [[] for _ in range(numCourses)]
        in_degree = [0] * numCourses
        
        for course, prereq in prerequisites:
            graph[prereq].append(course)  # prereq -> course
            in_degree[course] += 1
        
        # Start with all courses that have no prerequisites
        queue = deque([i for i in range(numCourses) if in_degree[i] == 0])
        completed = 0
        
        while queue:
            course = queue.popleft()
            completed += 1
            
            for next_course in graph[course]:
                in_degree[next_course] -= 1
                if in_degree[next_course] == 0:
                    queue.append(next_course)
        
        # If we completed all courses, there's no cycle
        return completed == numCourses
        